crypto/multisig/src/lib.rs

```diff
@@ -146,35 +187,47 @@ impl<P: Default + PublicKey + Message> SignerSet<P> {
         // While the verification code below should be immune to duplicate signers or
         // signatures, the overhead of deduping them is negligible and being
         // extra-safe is a good idea.
-        let mut potential_signers = self.signers.clone();
-        potential_signers.sort();
-        potential_signers.dedup();
+        let mut potential_individual_signers = self.individual_signers.clone();
+        potential_individual_signers.sort();
+        potential_individual_signers.dedup();
+
+        let mut potential_multi_signers = self.multi_signers.clone();
+        potential_multi_signers.sort();
+        potential_multi_signers.dedup();
 
         let mut signatures = multi_sig.signatures.clone();
         signatures.sort_by(|a, b| a.as_ref().cmp(b.as_ref()));
         signatures.dedup();
 
-        // See which signatures which match signers.
-        let mut matched_signers = Vec::new();
-        for signature in signatures.iter() {
-            let matched_signer = potential_signers.iter().find_map(|signer| {
-                signer
-                    .verify(message, signature)
-                    .ok()
-                    .map(|_| signer.clone())
-            });
-            if let Some(matched_signer) = matched_signer {
-                potential_signers.retain(|signer| signer != &matched_signer);
-                matched_signers.push(matched_signer);
+        // Collect individual signer identities that signed the message, and count how
+        // many signers we successfully matched.
+        let mut matched_signer_identities = Vec::new();
+        let mut num_matches = 0;
+
+        for individual_signer in potential_individual_signers {
+            if signatures
+                .iter()
+                .any(|sig| individual_signer.verify(message, sig).is_ok())
+            {
+                num_matches += 1;
+                matched_signer_identities.push(individual_signer);
+            }
+        }
+
+        // See if any multi-signer sets signed the message.
+        for signer_set in potential_multi_signers {
+            if let Ok(signer_identities) = signer_set.verify(message, multi_sig) {
+                matched_signer_identities.extend(signer_identities);
+                num_matches += 1;
             }
         }
 
         // Did we pass the threshold of verified signatures?
-        if matched_signers.len() < self.threshold as usize {
+        if num_matches < self.threshold as usize {
             return Err(SignatureError::new());
         }
 
-        Ok(matched_signers)
+        Ok(matched_signer_identities)
```

Comment:
:wrench: thinking about the common signer more, should this be sorted and deduped before returning?

Reply:
Correct, it should be. Good catch.